trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/collections/MultiMap.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.collections;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+/**
+ * A map between keys and values, each entry may have one or more values
+ * Null keys are not supported.
+ * The map may be configured to be case-insensitive at construction time, in which case
+ * all keys are lowercased.
+ */
+public interface MultiMap
+{
+    /**
+     * Check whether this map has case-sensitive keys
+     */
+    boolean isCaseSensitiveKeys();
+
+    /**
+     * Obtain a copy of the keys of the data in the map
+     *
+     * @return Immutable copy of the keys of the map
+     */
+    Set<String> keySet();
+
+    /**
+     * Obtain the set of entries in the map
+     */
+    Set<Map.Entry<String, List<String>>> entrySet();
+
+    /**
+     * Get all values associated to a given key in the map.
+     * Returns an empty list if the key is not present in the map.
+     *
+     * @param key key
+     * @return list of values, empty list if key not present
+     * @throws NullPointerException if key is null
+     */
+    List<String> get(String key);
+
+    /**
+     * Obtain the first value associated with a key, null if no values are associated
+     *
+     * @return first value associated with the key, null if none
+     * @throws NullPointerException if key is null
+     */
+    String getFirst(String key);
```

Comment:
wdyt of returning `Optional<String>`? We prefer to not leak possible NPEs.

Reply:
Happy to return an Optional, I was trying to stick to the same API as the other collections but there's no need!